### PR TITLE
deepsparse.Pipeline - generic pipeline, deepsparse.server support, NLP,IC,OD pipelines

### PR DESCRIPTION
**Pull request by bfineran**
co-authored by @rahul-tuli

The goal of this feature is to create a generic Pipeline class to abstract the general flow of wrapping `Engine` objects with pre/post processing and standardizing inputs/outputs with Pydantic models. These Pipelines should be easily dropped into deepsparse scripts for serving, inference, benchmarking, etc

instantiate a Pipeline using `Pipeline.create`
```python
from deepsparse import Pipeline

qa_pipeline = Pipeline.create("qa")
qa_pipeline(question="where do i live", context="i am an engineer and i live in america")
```

TODO:
- [x] Pipeline base class (#315)
- [x] PIpelineConfig (#318)
- [x] deepsparse.server refactor for Pipeline compatibility (#319)
- [x] deepsparse.transformers refactor to use Pipeline base class (#320)
- [x] deepsparse.transformers pipeline cleanup (#335)
- [x] support for sparsezoo model defaults included with `Pipeline.register` (#339)
- [x] rename _model to _schema (#340)
- [x] Image Classification Pipeline (#322)
- [x] Image Classification Pipeline scripts (#328)
- [x] YOLO Pipeline (#327)
- [x] YOLO annotation script refactor (#332)

### CODE REVIEW

File: src/deepsparse/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+
+
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Union
+
+import numpy
+from pydantic import BaseModel
+
+from deepsparse import Engine, Scheduler
+from deepsparse.benchmark import ORTEngine
```
Comment (rahul-tuli) [inline src/deepsparse/pipeline.py:28]: Should be changed to `deepsparse.benchmark_model`

```suggestion
from deepsparse.benchmark_model import ORTEngine
```

File: src/deepsparse/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    "ORT_ENGINE",
+    "SUPPORTED_PIPELINE_ENGINES",
+    "Pipeline",
+]
+
+
+DEEPSPARSE_ENGINE = "deepsparse"
+ORT_ENGINE = "onnxruntime"
+
+SUPPORTED_PIPELINE_ENGINES = [DEEPSPARSE_ENGINE, ORT_ENGINE]
```
Comment (rahul-tuli) [inline src/deepsparse/pipeline.py:43]: I think this should be refactored into an `Enum`, having both `ORT_ENGINE` and `ORTEngine` could potentially be confusing

**Review by dbogunowicz — COMMENTED** [review]
Neat. 

Silly question, but I think I have missed some context while on my short break: this object would be the template Pipeline for any future ML inference tasks derived from our SparseML pipelines?
i.e. we would use this object to build scripts to run deepsparse inference for IC, transformers, Unet, etc... is that right?

File: src/deepsparse/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+    """
+    Generic Pipeline abstract class meant to wrap inference engine objects to include
+    data pre/post-processing. Inputs and outputs of pipelines should be serialized
+    as pydantic Models.
+
+    Pipelines should not be instantiated by their constructors, but rather the
+    `Pipeline.create()` method. The task name given to `create` will be used to
+    load the appropriate pipeline. When creating a Pipeline, the pipeline should
+    inherit from `Pipeline` and implement the `setup_onnx_file_path`, `process_inputs`,
+    `process_engine_outputs`, `input_model`, and `output_model` abstract methods.
```
Comment (bfineran) [inline src/deepsparse/pipeline.py:59]: change `_model` to `Schema`

File: src/deepsparse/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+        if not isinstance(pipeline_inputs, self.input_model):
+            raise ValueError(
+                f"Calling {self.__class__} requires passing inputs as an "
+                f"{self.input_model} object or a list of kwargs used to create "
+                f"a {self.input_model} object"
+            )
+
+        # run pipeline
+        engine_inputs: List[numpy.ndarray] = self.process_inputs(pipeline_inputs)
+        engine_outputs: List[numpy.ndarray] = self.engine(engine_inputs)
```
[the remark this answers is not in the document]
Comment (bfineran) [reply src/deepsparse/pipeline.py:153]: yeah just a signal since it's not super obvious and needs to be in a specific format

**Review by markurtz — DISMISSED** [review]

**Review by markurtz — DISMISSED** [review]

**Review by KSGulin — DISMISSED** [review]
Approved pending fix for `pipeline.output_model` discussed previously

**Review by markurtz — APPROVED** [review]

**Review by KSGulin — APPROVED** [review]

**Pull request merged**